PurgeNugetFeeds.ps1 - clean up external feeds

geeeyetee opened the pull request:
.SYNOPSIS
Applies Octopus release retention policies to associated packages on NuGet feeds.

.DESCRIPTION
Mimics how Octopus natively applies retention policies to its Built-in Package Repository, but applies them to
external NuGet feeds. ref: https://octopus.com/docs/administration/retention-policies#built-in-repository

Uses Octopus and NuGet APIs to identify and delete any package that meets all of these criteria:
* the package is on an Octopus-configured External Feed that:
 * is of type Nuget Feed
 * has a feed username (API key) defined
 * meets the -ExcludeFeedRegex and -IncludeFeedRegex criteria
* the package's name is associated with an Octopus release
* the package's name/version combination is not associated with an Octopus release
* the package's version is lower than the highest version of the package associated with an Octopus release

.PARAMETER ExcludeFeedRegex
Packages will not be purged from feeds having URLs that match this regular expression. Examples:
-ExcludeFeedRegex '$^' # Default. Purge packages from any feeds with URLs that match -IncludeFeedRegex.
-ExcludeFeedRegex '/foo-core/' # Do not purge packages from e.g. feed http://foonugetserver/foo-core/nuget

.PARAMETER IncludeFeedRegex
Packages will not be purged from feeds having URLs that do not match this regular expression. Examples:
-IncludeFeedRegex '.*' # Default. Purge packages from any feeds except those with URLs that match -ExcludeFeedRegex.
-IncludeFeedRegex 'foo-hi' # Purge packages from e.g. feed http://foonugetserver/foo-hi/nuget

.PARAMETER OctopusUri
Example:
-OctopusUri 'http://foooctoserver/api'

.PARAMETER OctopusApiKey
Duh.

.PARAMETER NugetPath
Full path to the nuget CLI executable.

.PARAMETER PathToStoreDataAcrossRuns
Full path to a folder where files from the script run can be accessed by the subsequent script run. Example:
-PathToStoreDataAcrossRuns '\\foonas\fooshare\Octopus-PurgeNugetFeeds-Data'

.PARAMETER SpaceId
The space ID (not name). This only affects which NuGet feeds are queried for candidate packages to purge. Example:
-SpaceId 'Spaces-1' # Default

.PARAMETER ProceedEvenIfPreviousRunMayHavePurgedPackagesItShouldNotHave
If this switch is present, the script will ignore the result of the following check: At the beginning of each
run, the script saves a file under -PathToStoreDataAcrossRuns listing any packages that Octopus reports as
missing from feeds that comply with -ExcludeFeedRegex and -IncludeFeedRegex. Just before saving that file, the
script compares the current list with the one saved the previous time the script was run. If the current list
includes any packages that the previous list does not, the cause could be that during the
previous run the script did not work as intended and purged packages still associated with releases.
To avoid the possibility of further data loss, the script throws an exception and stops without purging
anything (unless this switch is present).

.PARAMETER UseCachedListOfPackagesInUse
Intended only to speed up debugging. Loads the list of packages associated with Octopus releases from a file
saved during the previous run instead of calculating the list by querying Octopus. Example:

.OUTPUTS

None

.EXAMPLE

Syntax for Arguments parameter of TFS Powershell step:
-Confirm:$($false) -WhatIf:$$(foo.WhatIf) -Verbose:$$(system.debug) -IncludeFeedRegex '/foo-' -ExcludeFeedRegex '/foo-core/' -OctopusUri '$(foo.OctopusUri)' -OctopusApiKey '$(foo.OctopusApiKey)' -PathToStoreDataAcrossRuns $(foo.PathToStoreDataAcrossRuns)

OctopusBob commented:
Please remove the .dlls from the commit.  It is okay to reference them in your script, just indicate to people where they can pull them.  This way users can pull the latest version of the .dlls without being tied to a specific version

geeeyetee commented:
@OctopusBob - I removed the .dlls and updated the script to automate pulling them down from nuget.org or any other configured/specified feeds. Thanks for the feedback!
